pybamm/solvers/idaklu_solver.py

```diff
@@ -44,6 +44,84 @@ def __init__(
 
         super().__init__("ida", rtol, atol, root_method, root_tol, max_steps)
 
+    def set_tolerances_by_variable(self, variables_with_tols, model):
+        """
+        A method to set the absolute tolerances in the solver by state variable.
+
+        Parameters
+        ----------
+        variables_with_tols : dict
+            A dictionary with keys that are strings indicating the variable you
+            wish to set the tolerance of and values that are the tolerances.
+
+        model : :class:`pybamm.BaseModel`
+            The model that is going to be solved.
+        """
+
+        size = model.concatenated_initial_conditions.size
+        self._check_atol_type(size)
+        for var, tol in variables_with_tols.items():
+            variable = model.variables[var]
+            if isinstance(variable, pybamm.StateVector):
+                self.set_state_vec_tol(variable, tol)
+            elif isinstance(variable, pybamm.Concatenation):
+                for child in variable.children:
+                    if isinstance(child, pybamm.StateVector):
+                        self.set_state_vec_tol(child, tol)
+                    else:
+                        raise pybamm.SolverError(
+                            """Can only set tolerances for state variables
+                            or concatenations of state variables"""
+                        )
+            else:
+                raise pybamm.SolverError(
+                    """Can only set tolerances for state variables or
+                    concatenations of state variables"""
+                )
+
+    def set_state_vec_tol(self, state_vec, tol):
+        """
+        A method to set the tolerances in the atol vector of a specific
+        state variable.
+
+        Parameters
+        ----------
+        state_vec : :class:`pybamm.StateVector`
+            The state vector to apply to the tolerance to
+        tol: float
+            The tolerance value
+        """
+        slices = state_vec.y_slices[0]
+        self._atol[slices] = tol
+
+    def _check_atol_type(self, size):
```

Comment:
make it clear in the name that this function also *changes* self._atol. In fact this could all go in `@atol.setter`, either for KLU or BaseSolver